pkg/indexer/blockTracker.go

```diff
@@ -0,0 +1,93 @@
+package indexer
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"sync/atomic"
+
+	"github.com/xmtp/xmtpd/pkg/db/queries"
+)
+
+/*
+*
+BlockTracker keeps a database record of the latest block that has been indexed for a contract address
+and allows the user to increase the value.
+*
+*/
+type BlockTracker struct {
+	latestBlock     atomic.Uint64
+	contractAddress string
+	queries         *queries.Queries
+	mu              sync.Mutex
+}
+
+// Return a new BlockTracker initialized to the latest block from the DB
+func NewBlockTracker(
+	ctx context.Context,
+	contractAddress string,
+	queries *queries.Queries,
+) (*BlockTracker, error) {
+	bt := &BlockTracker{
+		contractAddress: contractAddress,
+		queries:         queries,
+	}
+
+	latestBlock, err := getLatestBlock(ctx, contractAddress, queries)
+	if err != nil {
+		return nil, err
+	}
+	bt.latestBlock.Store(latestBlock)
+
+	return bt, nil
+}
+
+func (bt *BlockTracker) GetLatestBlock() uint64 {
+	return bt.latestBlock.Load()
+}
+
+func (bt *BlockTracker) UpdateLatestBlock(ctx context.Context, block uint64) error {
+	// Quick check without lock
+	if block <= bt.latestBlock.Load() {
+		return nil
+	}
+
+	bt.mu.Lock()
+	defer bt.mu.Unlock()
+
+	// Re-check after acquiring lock
+	if block <= bt.latestBlock.Load() {
+		return nil
+	}
+
+	if err := bt.updateDB(ctx, block); err != nil {
```

Comment:
I didn't want to over-think it for now, but I may end up debouncing updates to the DB so that it doesn't update on every block when you are doing a historical sync